aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/utils/SecureRandomSupplier.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazonaws.xray.utils;
+
+import java.lang.reflect.Field;
+import java.security.SecureRandom;
+import java.util.function.Supplier;
+
+/**
+ * An <em>internal</em> utility class to efficiently generate {@code SecureRandom} instances. This class
+ * should not be used outside of the AWS X-Ray SDK.
+ */
+public final class SecureRandomSupplier implements Supplier<SecureRandom> {
+    private static final Supplier<SecureRandom> RANDOM_SUPPLIER;
+    public static final SecureRandomSupplier INSTANCE = new SecureRandomSupplier();
+
+    static {
+        // JDK 9+: Use a thread-local SecureRandom instance if we're able to determine
```

Comment:
We removed threadlocal securerandom here after adding benchmarks for it in #162

I believe it's because the OS's random is guaranteed to be a global lock, so a thread-local securerandom instance doesn't reduce locking but does add overhead for the threadlocal lookup.

If this makes sense (maybe my benchmark is missing something, never know), could you revert this change?

Reply:
the entropy pool is just one source of blocking and actually isn't directly queried all that often (usually only for seed generation). the main source of contention I'm worried about is the thread-safety of the RNG algorithm itself – RNGs typically maintain internal state that prevent concurrent reads/updates. This blob of code checks for when that's true and uses thread locals as necessary.

what os/arch did you run your benchmarks on? what crypto provider(s) did you install? which JVM – corretto? [AmazonCorrettoCryptoProvider](https://aws.amazon.com/about-aws/whats-new/2019/07/introducing-the-amazon-corretto-crypto-provider/) provides a thread-safe `SecureRandom` implementation that stores its state in thread-locals, so `SecureRandom` skips synchronizing on calls into it. You'll see lock contention in SecureRandom if you use a crypto provider that doesn't provide a thread-safe random.

Reply:
now that I think about it, it would probably be better to always use a thread-local SecureRandom since "thread-safe" could also be implemented by just slapping `synchronized` on your random spi's methods.

Reply:
It sounds like coretto would be even faster without the threadlocal. But I just tested with normal openjdk 14 on Mac and even with that, not using threadlocal was a bit faster. I don't know how well my theory checks out, but since the benchmark numbers aren't convincing and threadlocal securerandom adds complexity, let's go ahead and stick with the singleton please. Thanks!